File: benchmark/synduce/list/issorted.f
Inductive SList = elt Int | cons {Int, SList};
Inductive CList = single Int | concat {CList, CList};

cat_list = fix (
  \f: SList -> SList -> SList. \xs: SList. \ys: SList.
  match xs with
    elt a -> cons {a, ys}
  | cons {a, b} -> cons {a, f b ys}
  end
);

repr = fix (
  \f: CList -> Compress SList. \xs: CList.
  match xs with
    single a -> elt a
  | concat {a, b} -> cat_list (f a) (f b)
  end
);

spec = \xs: SList. (fix (
  \f: SList -> {Int, Int, Bool}. \xs: SList.
  match xs with
    elt a -> {a, a, true}
  | cons {hd, tl} ->
    let result = f tl in
    {hd, result.2, and result.3 (< hd result.1)}
  end
) xs).3;

main = \xs: CList. spec (repr xs);

File: ml-benchmark/synduce/list/issorted.ml
type slist = Elt of int | Cons of int * slist
type clist = Single of int | Concat of clist * clist

let rec cat_list xs ys =
  match xs with
  | Elt a -> Cons (a, ys)
  | Cons (a, b) -> Cons (a, cat_list b ys)

let rec first = function 
  | Elt a -> a
  | Cons (a, _) -> a

let rec last = function 
  | Elt a -> a
  | Cons (_, t) -> last t

val repr: clist -> slist compress
let rec repr xs =
  match xs with
  | Single a -> Elt a
  | Concat (a, b) -> cat_list (repr a) (repr b)

let spec xs =
  let rec f xs =
    match xs with
    | Elt a -> (a, a, true)
    | Cons (hd, tl) ->
        let r = f tl in
        match r with
        | (r1, r2, r3) -> (hd, r2, r3 && hd < r1)
  in
  let t = f xs in
  match t with
  | (_, _, b) -> b

let program xs = spec (repr xs)